### ISSUE TITLE
Crash when Directive code is lexicalSuccessor of some evaluation.

### ISSUE THREAD

**Issue by mikaoP** [flang-compiler/f18-llvm-project#797]
In the following code:
```fortran
PROGRAM P
    IMPLICIT NONE
    INTEGER ::X

    IF (X /= 1) STOP 1

!    X = 44

    !$OMP PARALLEL
    X = 777
    !$OMP END PARALLEL

END PROGRAM P
```
The OpenMP directive seems to be not taken into account as a lexical successor of some evaluation when building PFT and I hit the assert
```
mlir/lib/IR/Block.cpp:217: mlir::Operation *mlir::Block::getTerminator(): Assertion `!empty() && back().mightHaveTrait<OpTrait::IsTerminator>()' failed.
```
I've managed to fix it by allowing directives as a lexicalSuccessor with:
```diff
diff --git a/flang/lib/Lower/PFTBuilder.cpp b/flang/lib/Lower/PFTBuilder.cpp
index 4a3f5bc05b45..8fdca653b879 100644
--- a/flang/lib/Lower/PFTBuilder.cpp
+++ b/flang/lib/Lower/PFTBuilder.cpp
@@ -371,7 +371,7 @@ private:
     auto &entryPointList = eval.getOwningProcedure()->entryPointList;
     evaluationListStack.back()->emplace_back(std::move(eval));
     lower::pft::Evaluation *p = &evaluationListStack.back()->back();
-    if (p->isActionStmt() || p->isConstructStmt() || p->isEndStmt()) {
+    if (p->isActionStmt() || p->isConstructStmt() || p->isEndStmt() || p->isDirective()) {
       if (lastLexicalEvaluation) {
         lastLexicalEvaluation->lexicalSuccessor = p;
         p->printIndex = lastLexicalEvaluation->printIndex + 1;
```
But I'm not sure about the implications of doing this.

EDIT: Compile command `bbc -fopenmp -emit-fir f1.f90`

**Comment by vdonaldson** [comment]
This is a case where structured code following unstructured code starts a new explicit basic block.  The IF statement is unstructured because the STOP statement is an unstructured branch.  The code that follows the IF statement is structured.  Like Construct PFT nodes, Directive PFT nodes contain nested Evaluations.  The first nested Evaluation may start a new block.

**Comment by vdonaldson** [comment]
Fix in PR #798
